### ISSUE TITLE
Possible another expat thing

### ISSUE THREAD

**Issue by robe2** [pramsey/pgsql-ogr-fdw#34]
I was trying a query with this:

http://download.geofabrik.de/north-america/us/district-of-columbia.html

```
 ogr_fdw_info -s "C:\fdw_data\gisdata\district-of-columbia-latest.osm" -l lines

```

which outputs this:

```
     CREATE SERVER myserver
      FOREIGN DATA WRAPPER ogr_fdw
      OPTIONS (
        datasource 'C:\fdw_data\gisdata\district-of-columbia-latest.osm',
        format 'OSM' );
    
    CREATE FOREIGN TABLE lines (
      fid integer,
      geom geometry,
      osm_id varchar,
      name varchar,
      highway varchar,
      waterway varchar,
      aerialway varchar,
      barrier varchar,
      man_made varchar,
      z_order integer,
      other_tags varchar )
      SERVER myserver
      OPTIONS ( layer 'lines' );

```

However when I go to query the table it created, it complains that z_order is a string and not an integer.

So I changed the definition to integer and notice, it seems the z_order is missing and what ends up showing in z_order is the other_tags value.

I checked to see if perhaps a gdal issue by running this query is is using the same gdal library as my ogr_fdw

```
ogrinfo "C:\fdw_data\gisdata\district-of-columbia-latest.osm" lines | more

```
And that looked right, z_order values were there and data structure agrees with ogr_fdw

```
Layer name: lines
Geometry: Line String
Feature Count: -1
Extent: (-77.120100, 38.791340) - (-76.909060, 38.996030)
Layer SRS WKT:
GEOGCS["WGS 84",
    DATUM["WGS_1984",
        SPHEROID["WGS 84",6378137,298.257223563,
            AUTHORITY["EPSG","7030"]],
        TOWGS84[0,0,0,0,0,0,0],
        AUTHORITY["EPSG","6326"]],
    PRIMEM["Greenwich",0,
        AUTHORITY["EPSG","8901"]],
    UNIT["degree",0.0174532925199433,
        AUTHORITY["EPSG","9108"]],
    AUTHORITY["EPSG","4326"]]
osm_id: String (0.0)
name: String (0.0)
highway: String (0.0)
waterway: String (0.0)
aerialway: String (0.0)
barrier: String (0.0)
man_made: String (0.0)
z_order: Integer (0.0)
other_tags: String (0.0)
OGRFeature(lines):4681186
  osm_id (String) = 4681186
  name (String) = (null)
  highway (String) = residential
  waterway (String) = (null)
  aerialway (String) = (null)
  barrier (String) = (null)
  man_made (String) = (null)
  z_order (Integer) = 3
  other_tags (String) = (null)
  LINESTRING (-77.1082606 38.935738,-77.1082557 38.9358361,-77.1082555 38.9359244,-77.1082552 38.9360971,-77.1082544 38.9365412,-77.1082541 38.9366631)

OGRFeature(lines):4725152
  osm_id (String) = 4725152
  name (String) = Texas Avenue Southeast
  highway (String) = secondary
  waterway (String) = (null)
  aerialway (String) = (null)
  barrier (String) = (null)
  man_made (String) = (null)
  z_order (Integer) = 6
  other_tags (String) = "HFCS"=>"Collector","lanes"=>"1","oneway"=>"yes","source:HFCS"=>"District of Columbia (DC GIS)"
  LINESTRING (-76.9386616 38.8891714,-76.9383773 38.8893008,-76.9382276 38.889401,-76.9381042 38.8895367,-76.938026 38.8896185,-76.9379174 38.8896617)
```

I don't seem to have the same issue if I use the osm.pbf version of the file.

I'm going to try next with a small .osm file to rule out its something specific to this file.

**Comment by robe2** [comment]
For completeness, I had to change my lines table in postgresql to this so I could select from it:

```
CREATE FOREIGN TABLE lines (
  fid integer,
  geom geometry,
  osm_id varchar,
  name varchar,
  highway varchar,
  waterway varchar,
  aerialway varchar,
  barrier varchar,
  man_made varchar,
  z_order varchar,
  other_tags varchar )
  SERVER myserver
  OPTIONS ( layer 'lines' );
```

and if I do a:

```
SELECT z_order, other_tags FROM lines WHERE osm_id = '4725152';
```
(takes about 21 seconds to return a notice: )

```
NOTICE:  handling more pgsql fields (10) than there are ogr fields (8)

NOTICE:  handling more pgsql fields (10) than there are ogr fields (8)
                                             z_order                                             | other_tags
-------------------------------------------------------------------------------------------------+------------
 "HFCS"=>"Collector","lanes"=>"1","oneway"=>"yes","source:HFCS"=>"District of Columbia (DC GIS)" |
(1 row)
```

**Comment by robe2** [comment]
Okay I created a small example which I generated with this command:

```
 BBOX="-122.8,45.4,-122.799,45.41"
wget --progress=dot:mega -O "small.osm" "http://www.overpass-api.de/api/xapi?*[bbox=${BBOX}][@meta]"

```

In this case the geometry is missing though the z-order is there and the ogrinfo looks okay. 

You can get the file from http://www.bostongis.com/postgisstuff/small.osm.zip

But I think I'm going nuts, because I just retried the dc file and that one seems fine now.

**Comment by robe2** [comment]
To add to the confusion to this.  I walked away from my computer for a couple of hours.  Had my windows still open when I came back and decided to try the queries again with my already registered fdw tables that were having the issue.  Both the small.osm and the osm dc I was having trouble with all worked now.

So if there is an issue here it's not consistently replicable.  FWIW I'm running the GDAL 2.0.1 patched version (which I compiled a little after Even made the fix for Expat)  in #29

**Comment by robe2** [comment]
More craziness,  @pramsey  do you see the z_order field in your output?

This is beginning to sound like a gdal bug or a bug in my head.  I thought I was only having issue with o.osm but same thing happens with .pbf.

Anyway I discovered I never have this issue on my production server, but to my amazement the same set of files do not output the same structure on my windows 7 64-bit desktop as they do on my windows 2012 64-bit server and they are running the same binaries for gdal and ogrfdw.

The server generates this -- no z_order field. (and ogr2ogr I run on the server, also doesn't have that z_order field)

```
CREATE SERVER myserver
  FOREIGN DATA WRAPPER ogr_fdw
  OPTIONS (
    datasource 'c:\fdw_data\gisdata\small.osm',
    format 'OSM' );

CREATE FOREIGN TABLE lines (
  fid integer,
  geom geometry,
  osm_id varchar,
  name varchar,
  highway varchar,
  waterway varchar,
  aerialway varchar,
  barrier varchar,
  man_made varchar,
  other_tags varchar )
  SERVER myserver
  OPTIONS ( layer 'lines' );

```

Run the same on my windows 7 desktop  (I checked twice that they are running the same dated binaries) and I get this:

```
CREATE SERVER myserver
  FOREIGN DATA WRAPPER ogr_fdw
  OPTIONS (
    datasource 'c:\fdw_data\gisdata\small.osm',
    format 'OSM' );

CREATE FOREIGN TABLE lines (
  fid integer,
  geom geometry,
  osm_id varchar,
  name varchar,
  highway varchar,
  waterway varchar,
  aerialway varchar,
  barrier varchar,
  man_made varchar,
  z_order integer,
  other_tags varchar )
  SERVER myserver
  OPTIONS ( layer 'lines' );
```

Not sure how the same exact file can output two different set of fields.  Anyway my desktop seems okay now, magically, for now and my queries and ogr2ogr agree there is a z_order field there and are outputting it. Why my server doesn't see that field and doesn't care is a mystery.

**Comment by robe2** [comment]
Okay I think I figured out the problem and it's not a problem, just user misunderstanding.  

Turns out the z_order field output is controlled by https://svn.osgeo.org/gdal/trunk/gdal/data/osmconf.ini  (which is stored in the gdal-data folder).  I think my production server is still using the 1.11.1 version, which maybe is set to not output z_order.  So though I am running the same binaries, I suspect my osmconf.ini is configured differently.

and this new z_order feature computation came in GDAL 2.0 according to this:
http://www.gdal.org/drv_osm.html
```
tarting with GDAL 2.0, fields can be computed with SQL expressions (evaluated by SQLite engine) from other fields/tags. For example to compute the z_order attribute.

```

Why my desktop seemed to show an issue is probably cause I hit the wrong batch script when launching my postgres instance.  I have one that sets the GDAL env and one that doesn't. Too complicated to get into.

So I'm going to just close this out as user ignorance.
